### ISSUE TITLE
touching jog wheel causes too much adjustment, waveform out of sync with audio

### ISSUE THREAD

**Issue by ywwg** [mixxxdj/mixxx#11711]
### Bug Description

A couple times during practice, when I touch the jog wheel to adjust the sync (my controller uses the "jog" parameter), a weird thing happens:

1. the track speed changes MUCH more than it should.  the audio bends noticeably. 
2. But, the waveform display does not seem to move a lot! In fact, the audio becomes out of sync with the waveform display. The beats are noticeably misaligned from the waveform and beat display.

If I pause and play, the waveform/audio sync is restored.  Nothing else about the waveform display looks incorrect, it's just in the wrong place.

### Version

HEAD

### OS

Ubuntu 22.04

**Comment by ywwg** [comment]
I will try disabling the new opengl stuff and see if that does anything -- I don't expect it to

**Comment by ywwg** [comment]
I suspect this is a bug with the Traktor S3 script.  Merging the rewrite will likely fix it

**Comment by tekhedd** [comment]
I was worried it was just me, or something with my VCI-400 script. Since the wheel support is *very* generic, I don't see how that could happen. I thought it was my hardware so I *bought a new laptop*, did not fix it.

Occurs even if "vinyl" is turned off, it's not just me brushing the touch sensor.

I have been noticing the same thing, since about mixxx 2.3.8 or 2.3.6 (not sure, one of the later updates to 2.3). It also definitely happens with 2.4.1 on the latest ubuntu releases (lubuntu LTS). 

It doesn't seem to be related to the output audio or audio buffer. Usually happens when I have two (or more) decks going, and when I go to gently adjust the timing of one of the decks instead of getting a tiny pitch bend I get a 2+beat random skip. Sometimes the track display is still correct, sometimes it becomes offset from the audio and so I have to correct and beat match by ear from then on.

It is difficult to reproduce and intermittent. I suspect it has to do with midi overload. I will update my jog script to log any "large offsets" and see if I can track it downn.

**Comment by ywwg** [comment]
I have not experienced this issue in a long time fwiw

**Comment by tekhedd** [comment]
I can see why most people don't have the issue. It appears to be I/O related, and the kind of thing that you just go "well I'll try another port/computer". I have had intermittent (and frustrating) experiences with this issue for the last several months, but now I have a setup where it is repeatable, so I am hoping to at least know the source of the issue before moving on. :)

It seems like I am getting duplicated USB MIDI packets for the jog wheel (or whatever), and this seems to happen most often after seeking, which is to say during buffer-feeder I/O. My first clue was that the entire UI would freeze for a few seconds when I moved a jog wheel and also I could see from the output that it buffer feeder was working, usually after seeking. This behavior became *different* when I switched timing to the network adapter, and instead I got weird slowdowns, so clearly the audio clock freezes during this. (Hmm, typing this makes me want to try a different cable and or port).

Regardless of the source, the **failure mode** is what I'm actually interested in, here. When something goes wrong, it would be nice to fail gracefully...

I wrote some tracking code to watch the accumulated jog input, and it looks like after one of these "backups" there will be a sudden, large jog offset of >7000, even though I barely touched the wheel. Even at a multiplier of ~0.1 is still a big instantaneous jog. I'm wondering if an enormous, sudden jog could actually result in the display and beat timers being wrong? Yes, the offset will all be in tiny increments well within the spec of the "jog" parameter, but... 7000/64 means it called  setValue(...,'jog',-6.8) 109 times, and probably all within one tick of the clock timer, effectively an instantenous jog of -700 or so...

This is why I suspect packet duplication. Hmm.

**Comment by tekhedd** [comment]
It's GL. With a non-GL waveform I can not reproduce the random offsets. Also, the waveform display jumps around and flickers like a glitch video. It's fascinating and definitely not desirable. :}

I will chase specifics and update this comment. I need to confirm but I am pretty sure my old laptop also used GL, just different video drivers etc, for slightly more consistent results.

**Comment by benselme** [comment]
@tekhedd  Thanks a lot, your comments have been very helpful to me, especially the keylock trick that seems to be a reliable workaround. This problem has been plaguing me for a few weeks, ruining some recordings. Luckily, I don't play live at the moment.

So here are my findings so far:

1. It happens with both GL and non-GL waveforms
2. It happens when nudging without a controller (keyboard shortcuts in my case)
3. It has to do with keylock/transpose because:
    - it never happens with keylock on as far as I can tell
    - it nearly always happen if I nudge and nudge and nudge after transposing (either syncing or resetting key)
4. Sometimes the "big slip" happens (all of a sudden the nudge accelerates), and then it becomes impossible to beatmatch again (waveform alignment not matching sound, and erratic nudge control)
5. Also sometimes, I get the same symptoms with no "big slip".
7. It seems that some tracks are more vulnerable than others, at least I can reproduce the bug consistently with some specific pairs of tracks.

As for the environment, I'm on the latest Linux Mint on a Dell Latitude 5520, 1x Intel Core i5 Quad (i5-1145G7) 2.60 GHz, 16 GB RAM. Sound output is a Traktor Z1's. Nothing much installed, it's a computer I bought specifically to run Mixxx.

**Comment by tekhedd** [comment]
> [@tekhedd](https://github.com/tekhedd) Thanks a lot, your comments have been very helpful to me, especially the keylock trick that seems to be a reliable workaround. This problem has been plaguing me for a few weeks, ruining some recordings. Luckily, I don't play live at the moment.
>     2. It happens when nudging without a controller (keyboard shortcuts in my case)

Glad it helps someone! Thanks for saying something. I was starting to think I was all alone. 

Haven't tried to reproduce it on windows or mac, would be curious whether it's in the engine or the platform support... If it varies per-file (I believe it does for me too), maybe some sort of timing related issue in the decompression pipeline?

**Comment by ronso0** [comment]
@ywwg can you still reproduce it with the S3, or with steps @benselme posted?

**Comment by ywwg** [comment]
as stated above, I have not been able to reproduce this problem in a long time

**Comment by mxmilkiib** [comment]
maybe this is resolved?

**Comment by tekhedd** [comment]
I'm still seeing it on the latest stable deb style release. Give me a moment I'll go check it out.

er, I was able to get it (1.3.5, old rubberband, lubuntu) to glitch out within a few minutes, but it's always possible that was genuinely me. I'll spend some time with it this evening, and report in a little more detail, but at this moment the dog absolutely requires my full attention. :)

I have been using key lock to work around the problem, so I'd like to reproduce it a few more times before I claim it's still bugged. (It's so much more fun without key lock, although I suppose these days who besides me even does this?)

**Comment by tekhedd** [comment]
So, I was able in about 10 minutes to get deck 2 off by about a quarter of a beat.  (see picture) This is more or less perfectly beatmatched in the headphones.  Quickly pause/unpausing deck 2 restores the visual/audio sync.  This was accomplished with Vinyl mode disabled, there is no way I could have grabbed the decks. I have reproduced this problem with several different brands of controllers, using a mix of controller scripts written by me and stock ones, on two different computers, both running debian-based distros, with stock ASIO USB drivers. This system has a low latency kernel.

![Image](https://github.com/user-attachments/assets/d1bd33c9-868b-44ec-915e-c130ba272b8d)

The difference between this and previous versions is that it seems to smoothly stretch to the new offset instead of audibly jumping. Is there a prerelease version I should be testing that has significant changes that might affect this?

I am still not certain the exact sequence of events that is necessary to trigger an offset. I believe it is some combination of Change Pitch + Change Speed with the track running, followed by moving the jog wheel any amount. It may be that the Sync button triggers it. When I am doing mix-in-key it is normal for both tracks to always have some pitch and speed offset in a random direction, so it is difficult to isolate the specific circumstances.

I will do some tests to try to narrow it down. "No sync" session, "pitch and speed on only deck b" etc. Actually that sounds kind of fun.

Also thanks for giving me an excuse to blow off work for half an hour.

**Comment by ronso0** [comment]
> both running debian-based distros, with stock ASIO USB drivers.

Hmm? Why ASIO on Linux?

**Comment by tekhedd** [comment]
> > both running debian-based distros, with stock ASIO USB drivers.
> 
> Hmm? Why ASIO on Linux?

ROFL sorry I mean ALSA. Too many years doing audio development and the wrong acronym comes out of my fingers automatically. :D

**Comment by ronso0** [comment]
So this is still about the audio / visual offset, right?
Not just about some general drift due to controller wheels sending spontaneous signals.

**Comment by tekhedd** [comment]
Yes. The waveforms are not in sync with the playing audio. Additionally, I can confirm that the "beat" is not in sync with playing audio, because controller mappings I wrote flash leds on the downbeat, and those flashes are in sync with the beat as drawn on the screen, not the playing audio.

Sometimes this manifests as the audio suddenly jogging to a new location when you first barely touch the jog wheel. So, for example, when it is in this state, you will move the jog 1 tick but it will move the audio half a beat. This used to be a jump, but in recent versions it is smoothed.

I have reproduced the problem with various controllers. The code for this controller is pretty generic:

    Deck.prototype.handleWheel = function(channel, control, value, status, group) {
        var jogValue = value - 0x40; // Control centers on 0x40 (64)
    
        if (engine.isScratching(this.deckNumber)) {
            engine.scratchTick(this.deckNumber, jogValue);    // scratch
        }
        else {
            // Correct for high-res wheel (only if not scratching, which is 
            // corrected in the scratchEnble function)
    
            jogValue = jogValue * JOG_CORRECTION;
            engine.setValue(this.group, 'jog', jogValue);          // Pitch bend (or cue)
        }
    };

**Comment by ywwg** [comment]
Does this ever happen with keylock off or not adjusting musical pitch? (ie, pure linear stretcher only instead of rubberband)

**Comment by ywwg** [comment]
I suspect this new problem is not the same as my original bug report

**Comment by tekhedd** [comment]
1) yes this probably is not related to the original bug report, although it does look like one other person has had the problem.


2) This ONLY happens when two things are combined:

 a) keylock is OFF
 b) pitch is adjusted (and probably also speed, but speed is always adjusted because DJ, right?)

So, yes, this appears to have something to do with tracking changes in speed when the pitch has been adjusted *and* keylock is off. 

I suspect very few mix-in-key people also turn off keylock. I personally find it retains the organic feel of an unlocked mix but you still get the magic of "wow everything magically works together", but if the two people in this thread are the only people in the world that use it, maybe it's not worth fixing. I'm a programmer; I get it. :(

**Comment by benselme** [comment]
Yes, just like @tekhedd, as far as I can tell, it's bug that's specific to the combination of no keylock, speed AND key adjusted. A combination that actually doesn't make a lot of sense, IMHO. I haven't encountered that bug since I stopped mixing in key altogether (in part after encountering https://github.com/mixxxdj/mixxx/issues/12461)

**Comment by ywwg** [comment]
when keylock is off but pitch and speed are adjusted, we get a combination of two adjustments: key *and* speed. This combo could be causing the issue.  Does this happen with rubberband v2 and v3 as well?

**Comment by tekhedd** [comment]
> when keylock is off but pitch and speed are adjusted, we get a combination of two adjustments: key _and_ speed. This combo could be causing the issue. Does this happen with rubberband v2 and v3 as well?

Did v2 get replaced by v3? I just reproduced it with what is called "Rubberband" in the current release, which I assume is Rubberband v1. I previously reproduced it with "both versions" but I have no idea whether v2 was the other version at that time.

**Comment by ronso0** [comment]
FWIW yesterday I switched from RubberBand to R3 mid-session and (in addition to the sudden pitch change) the waveform was off by ~1 beat, like the audio / visual mismatch that's been reported lately.
Ejecting and reloading the track fixed it.

**Comment by ywwg** [comment]
> Did v2 get replaced by v3? I just reproduced it with what is called "Rubberband" in the current release, which I assume is Rubberband v1. I previously reproduced it with "both versions" but I have no idea whether v2 was the other version at that time.

I believe if you don't have R3 installed it won't appear in the list. default is v2

**Comment by tekhedd** [comment]
[edit] Well that's not confusing or anything. I see librubberband3 is at v4, and is tagged experimental, so I need to unmask it. Will try that tonight.

[edit2] Have built librubberband from source, 4.0.0 release, using the configuration "fftw/libsamplerate", as the docs recommend this for distribution (and the sid build appears to require libfftw and libsamplerate). Mixxx finds it just fine with LD_LIBRARY_PATH...so now I just need some free time for testing.



> > Did v2 get replaced by v3? I just reproduced it with what is called "Rubberband" in the current release, which I assume is Rubberband v1. I previously reproduced it with "both versions" but I have no idea whether v2 was the other version at that time.
> 
> I believe if you don't have R3 installed it won't appear in the list. default is v2

Ah, I see. [edit No, he didn't see.] Ubuntu is at version 3 of librubberband2. That's not confusing or anything. (So, when you say "Rubberband 3", on debian based distros this is librubberband2 v3, as all releases are binary compatible with v1 so there is no sane reason for the "2" in the name other than that there must have been a previous incompatible librubberband and why is everything always complicated?)

Um, my system has librubberband2 at 3.3.0. apt says this was already installed.

Hmm, and today when I launch Mixxx I see "R3" in the popup.  I've stay generally up to date (usually not more than a few weeks out of date), so it's doubtful that the version of rubberband is the issue, but I'll make a point to do more testing asap.

**Comment by ywwg** [comment]
thanks! yeah rubberband does not have an intuitive versioning scheme.  "R2" and "R3" are not version numbers, but refer to the engines you can choose from.  From their readme:

> The most important are the options -2 and -3. These select between two different processing engines, known as the R2 (Faster) engine and the R3 (Finer) engine. The R3 engine produces higher-quality results than R2 for most material, especially complex mixes, vocals and other sounds that have soft onsets and smooth pitch changes, and music with substantial bass content. However, it uses much more CPU than the R2 engine.
> 
> The R2 engine was the only method available in Rubber Band Library up to versions 2.x, and for compatibility it remains the default (in the case that neither -2 nor -3 is requested explicitly) whenever the command-line tool is invoked as rubberband. The R3 engine is the default if the tool is invoked as rubberband-r3.

**Comment by ywwg** [comment]
So with old version of rubberband, the R3 engine is not available

**Comment by tekhedd** [comment]
Hmm. Well, my popup says R3, with the system supplied librubberband2.so.2.3.0 (that is actually built from librubberband v3.3.0). So apparently LUbuntu *is* supplying R3 but I have no idea when/if that was added. 

So, I'll just test with the hot and fresh 4.0.0 build and eliminate all of the uncertainty. :D Fine, I didn't want to fight with this XP installer right now anyway, back in a few.

[edit] Inconclusive. I absolutely had an offset on the first or second track I tried, and then could not reproduce it for some time, and had to go back to work. Will spend quality time with it over the weekend.

[edit 2] OK I must be getting old and/or tired.  I had keylock on the whole time. Since the problem only occurs when keylock is off, no wonder I couldn't reproduce. I'll try again.

**Comment by tekhedd** [comment]
Was able to reproduce it in seconds, with librubberband 3 v4.0.0 (R3).

Will experiment with other algorithms (builtin etc) to see if it's algorithm related. 

I also note that I had not done a setcap, so the engine was not running with sched_fifo priority[*], and that I get audible glitching on bends when it goes off with R3. CPU meter shows plenty of headroom, but maybe connected to underflows or similar while bending? Hmm.

* - I've never in my life had an audio program that ran at lower latency with some or all of its threads priority boosted, on any platform.  That's kind of neat, actually. :)

**Comment by ywwg** [comment]
ok I will rename the Issue with the suspected cause

**Comment by tekhedd** [comment]
Cool! Note that this also does also happen with R2 and R1, but this does clarify that it happens *now*. Possibly the engine coders did not anticipate the "change pitch and speed in varying directions at once" use case.

Would it be useful to test if it happens *without* rubberband?

**Comment by benselme** [comment]
It's also been happening to me lately although I stopped using key lock or transpose altogether. Unfortunately I don't have yet a sure way to reproduce it but each time it occurs while looping and when I beat jump to the marker at the beginning of the loop. I then see this «gliding» effect in the waveform and I have to stop the track, turn sync off and on again. Sorry if this isn't terribly helpful!

**Comment by ywwg** [comment]
if you can record a screencast of the problem happening, that would be helpful!  My current hunch is that rubberband has an internal buffer that is getting much too big and is messing up the latency

**Comment by tekhedd** [comment]
I'll see what I can do. Never done a video capture from Linux before, how hard can it be? ;)  

Also, I might just do a build from source and see about chasing down the problem myself. After all, if I can reliably reproduce it, I should be debugging. I should probably start by building the project.

**Comment by ywwg** [comment]
I think GNOME and KDE both have built-in screencapture support, and there are good third party tools if you don't use those.

**Comment by tekhedd** [comment]
Hah, Gnome and KDE... we'll see how that translates in LXQt land. :) Been busy this winter, I swear I'll get to it. Clearly it's not a widespread issue so no sense of urgency LOL. 

In the meantime, I've been using good old Soundtouch and absolutely can not reproduce the problem unless I switch back to Rubberband. Its either a problem in the Rubberband library itself or in the Mixxx Rubberband-specific code which I imagine must be quite minimal. Wild.
